budgetportal/tests/test_prov_infra_chart.py

```diff
@@ -0,0 +1,846 @@
+        """Test that Q1, Q2 and Q3 are updated correctly when Q4 inserted"""
+        snapshots_data = time_series_data(self.project.project_snapshots.all())
+        snapshots_data = snapshots_data[u"snapshots"]
+        self.assertEqual(len(snapshots_data), 3)
+
+        # Check Q1 values
+        self.assertEqual(snapshots_data[0]["total_spent_in_quarter"], 12)
+        self.assertEqual(snapshots_data[0]["total_spent_to_date"], 212)
+
+        # Check Q2 values
+        self.assertEqual(snapshots_data[1]["total_spent_in_quarter"], 21)
+        self.assertEqual(snapshots_data[1]["total_spent_to_date"], 233)
+
+        # Check Q3 values
+        self.assertEqual(snapshots_data[2]["total_spent_in_quarter"], 30)
+        self.assertEqual(snapshots_data[2]["total_spent_to_date"], 263)
+
+        # Create Q4 snapshot
+        q4 = Quarter.objects.create(number=4)
+        irm_snapshot_4 = IRMSnapshot.objects.create(
+            financial_year=self.fin_year,
+            quarter=q4,
+            date_taken=date(2031, 3, 31),
+            file=File(self.file_4),
+        )
+        ProvInfraProjectSnapshot.objects.create(
+            irm_snapshot=irm_snapshot_4,
+            project=self.project,
+            actual_expenditure_q1=13,
+            actual_expenditure_q2=22,
+            actual_expenditure_q3=31,
+            actual_expenditure_q4=40,
+            expenditure_from_previous_years_total=200,
+        )
+        # Recreate the chart data
+        snapshots_data = time_series_data(self.project.project_snapshots.all())
+        snapshots_data = snapshots_data[u"snapshots"]
+        self.assertEqual(len(snapshots_data), 4)
+
+        # Check Q1 values
+        self.assertEqual(snapshots_data[0]["total_spent_in_quarter"], 13)
+        self.assertEqual(snapshots_data[0]["total_spent_to_date"], 213)
+
+        # Check Q2 values
+        self.assertEqual(snapshots_data[1]["total_spent_in_quarter"], 22)
+        self.assertEqual(snapshots_data[1]["total_spent_to_date"], 235)
+
+        # Check Q3 values
+        self.assertEqual(snapshots_data[2]["total_spent_in_quarter"], 31)
+        self.assertEqual(snapshots_data[2]["total_spent_to_date"], 266)
+
+
+class NullQuarterlySpendTestCase(TransactionTestCase):
+    def setUp(self):
+        self.file = open(EMPTY_FILE_PATH)
+        self.file_2 = open(EMPTY_FILE_PATH)
+        self.project = ProvInfraProject.objects.create(IRM_project_id=1)
+        self.fin_year = FinancialYear.objects.create(slug="2030-31")
+        q1 = Quarter.objects.create(number=1)
+        irm_snapshot_1 = IRMSnapshot.objects.create(
+            financial_year=self.fin_year,
+            quarter=q1,
+            date_taken=date(2030, 6, 30),
+            file=File(self.file),
+        )
+        ProvInfraProjectSnapshot.objects.create(
+            irm_snapshot=irm_snapshot_1,
+            project=self.project,
+            actual_expenditure_q1=None,
+            expenditure_from_previous_years_total=200,
+        )
```

Comment:
I think you only need the second snapshot for this test - we're testing subsequent nulls, not updates.

In fact, I think we only need one test case, not two: a Q3 or Q4 snapshot where Q1 and Q3 actuals are set and Q2 actual is null - it can show that Q1 spend to date is calculated and the rest are null.

Reply:
https://github.com/vulekamali/datamanager/pull/215/files/8003845d3d85a9e61ca08ad7c6fc0f02fe05b219#diff-93488925aca732879663affbdf6b3505